File: internal/oura/endpoints.go
package oura

const baseURL = "https://api.ouraring.com"

// EndpointSpec describes a single Oura API v2 endpoint.
type EndpointSpec struct {
	Name     string // e.g. "daily_sleep"
	Path     string // e.g. "/v2/usercollection/daily_sleep"
	HasDates bool   // supports start_date/end_date query params
	IsList   bool   // returns paginated list vs single object
	IDField  string // JSON field to use as oura_id ("id", "" for none)
	DayField string // JSON field to use as day ("day", "timestamp", "")
}

// Registry contains all Oura API v2 endpoints.
var Registry = []EndpointSpec{
	{Name: "daily_sleep", Path: "/v2/usercollection/daily_sleep", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "sleep", Path: "/v2/usercollection/sleep", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "sleep_time", Path: "/v2/usercollection/sleep_time", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "daily_activity", Path: "/v2/usercollection/daily_activity", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "daily_readiness", Path: "/v2/usercollection/daily_readiness", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "heartrate", Path: "/v2/usercollection/heartrate", HasDates: true, IsList: true, IDField: "", DayField: "timestamp"},
	{Name: "daily_resilience", Path: "/v2/usercollection/daily_resilience", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "daily_stress", Path: "/v2/usercollection/daily_stress", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "daily_spo2", Path: "/v2/usercollection/daily_spo2", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "daily_cardiovascular_age", Path: "/v2/usercollection/daily_cardiovascular_age", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "vo2_max", Path: "/v2/usercollection/vo2_max", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "workout", Path: "/v2/usercollection/workout", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "session", Path: "/v2/usercollection/session", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "tag", Path: "/v2/usercollection/tag", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "enhanced_tag", Path: "/v2/usercollection/enhanced_tag", HasDates: true, IsList: true, IDField: "id", DayField: "day"},
	{Name: "ring_configuration", Path: "/v2/usercollection/ring_configuration", HasDates: false, IsList: true, IDField: "id", DayField: ""},
	{Name: "rest_mode_period", Path: "/v2/usercollection/rest_mode_period", HasDates: true, IsList: true, IDField: "id", DayField: "start_day"},
	{Name: "personal_info", Path: "/v2/usercollection/personal_info", HasDates: false, IsList: false, IDField: "", DayField: ""},
}

// RegistryMap provides lookup by endpoint name.
var RegistryMap = func() map[string]EndpointSpec {
	m := make(map[string]EndpointSpec, len(Registry))
	for _, spec := range Registry {
		m[spec.Name] = spec
	}
	return m
}()

// EndpointNames returns all endpoint names.
func EndpointNames() []string {
	names := make([]string, len(Registry))
	for i, spec := range Registry {
		names[i] = spec.Name
	}
	return names
}

File: clients/mcp/src/oura_mcp/endpoints.py
"""Mirror of internal/oura/endpoints.go — the subset of fields the MCP uses.

Parity with the Go Registry is enforced by `make check-endpoints`. Only these
four fields matter to the MCP surface:
  - name: the endpoint identifier (used in URLs and tool names)
  - has_dates: whether the tool exposes start_date/end_date params
  - is_list: whether the tool exposes a limit param (vs. single-object)
  - has_day_field: whether the Go side indexes by day (informational here;
    surfaces via the parity check so drift fails fast)
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class EndpointSpec:
    name: str
    has_dates: bool
    is_list: bool
    has_day_field: bool


REGISTRY: tuple[EndpointSpec, ...] = (
    EndpointSpec("daily_sleep",              has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("sleep",                    has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("sleep_time",               has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("daily_activity",           has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("daily_readiness",          has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("heartrate",                has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("daily_resilience",         has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("daily_stress",             has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("daily_spo2",               has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("daily_cardiovascular_age", has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("vo2_max",                  has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("workout",                  has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("session",                  has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("tag",                      has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("enhanced_tag",             has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("ring_configuration",       has_dates=False, is_list=True,  has_day_field=False),
    EndpointSpec("rest_mode_period",         has_dates=True,  is_list=True,  has_day_field=True),
    EndpointSpec("personal_info",            has_dates=False, is_list=False, has_day_field=False),
)

REGISTRY_BY_NAME: dict[str, EndpointSpec] = {spec.name: spec for spec in REGISTRY}
ENDPOINT_NAMES: tuple[str, ...] = tuple(spec.name for spec in REGISTRY)
